delivery_carrier_return_barcode_pattern/models/delivery_carrier.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import re
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+HELP_RETURN_BARCODE_PATTERN = """
+    This pattern can be used to parse a return barcode from the carrier.
+    This has to be a valid regular expression.
+    See https://docs.python.org/3/library/re.html
+"""
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    return_barcode_pattern = fields.Char(help=HELP_RETURN_BARCODE_PATTERN)
+
+    @api.constrains("return_barcode_pattern")
+    def _constrains_return_barcode_pattern(self):
+        for record in self:
+            if record.return_barcode_pattern:
+                try:
+                    re.compile(record.return_barcode_pattern)
+                except (re.error):
+                    raise ValidationError(
+                        _("Return Barcode Pattern must be a valid regular expression")
+                    )
+
+    @api.model
+    def _get_origin_from_barcode(self, barcode):
+        """Return a list of strings, parsed with carrier return barcode patterns.
+
+        Even if no pattern matched the barcode, it will return a list containing
+        the barcode itself.
+        """
+        pattern_query = """
+            SELECT DISTINCT(COALESCE(return_barcode_pattern, '(.*)'))
+            FROM delivery_carrier;
+        """
+        self.flush()
+        self.env.cr.execute(pattern_query)
+        patterns = [row[0] for row in self.env.cr.fetchall()]
+        res = []
+        for pattern in patterns:
+            match = re.match(pattern, barcode)
+            if match:
+                res.append(match.groups()[0])
```

Comment:
The pattern could contain several information that can be useful. I would split this method: one method to load the matches, and one method to return all origins from all the matches.

With such regex `Y(?P<origin>.*)` you can type your groups.

```suggestion
                res.append(match.group("origin"))
```